File: artifacts/tulipindicators/decompose/src/apo.c
int ti_apo_start(const double *options)
{
  (void) options;
  return 1;
}


/*** DEPENDENCIES:
***/


int ti_apo(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *input = inputs[0];
  double *apo = outputs[0];
  unsigned int apo_idx = 0;
  const int short_period = (int) options[0];
  const int long_period = (int) options[1];
  if (short_period < 1)
  {
    return 1;
  }
  if (long_period < 2)
  {
    return 1;
  }
  if (long_period < short_period)
  {
    return 1;
  }
  if (size <= ti_apo_start(options))
  {
    return 0;
  }
  double short_per = 2 / (((double) short_period) + 1);
  double long_per = 2 / (((double) long_period) + 1);
  double short_ema = input[0];
  double long_ema = input[0];
  int i;
  for (i = 1; i < size; i += 1)
  {
    short_ema = ((input[i] - short_ema) * short_per) + short_ema;
    long_ema = ((input[i] - long_ema) * long_per) + long_ema;
    const double out = short_ema - long_ema;
    apo[apo_idx] = out;
    apo_idx += 1;
  }

  assert(((&apo[apo_idx]) - outputs[0]) == (size - ti_apo_start(options)));
  return 0;
}


/*** DEPENDENCIES:
int ti_apo_start(const double *options)
{
  (void) options;
  return 1;
}


----------------------------
***/

File: artifacts/tulipindicators/translate/src/apo.rs
use crate::*;
use std::assert;

pub fn ti_apo_start(options: Option<&[f64]>) -> i32 {
    let _ = options; // Explicitly ignore the parameter to match C behavior
    1
}

pub fn ti_apo(
    size: i32,
    inputs: Option<&[&[f64]]>,
    options: Option<&[f64]>,
    outputs: Option<&mut [&mut [f64]]>,
) -> i32 {
    // Check for None (equivalent to NULL checks in C)
    if inputs.is_none() || options.is_none() || outputs.is_none() {
        return 1;
    }

    let inputs = inputs.unwrap();
    let options = options.unwrap();
    let outputs = outputs.unwrap();

    // Get input and output slices
    let input = inputs[0];
    let apo = &mut outputs[0];

    let short_period = options[0] as i32;
    let long_period = options[1] as i32;

    // Validation checks
    if short_period < 1 || long_period < 2 || long_period < short_period {
        return 1;
    }

    let start = ti_apo_start(Some(options));
    if size <= start {
        return 0;
    }

    let short_per = 2.0 / (short_period as f64 + 1.0);
    let long_per = 2.0 / (long_period as f64 + 1.0);

    let mut short_ema = input[0];
    let mut long_ema = input[0];
    let mut apo_idx = 0;

    for i in 1..size as usize {
        short_ema = (input[i] - short_ema) * short_per + short_ema;
        long_ema = (input[i] - long_ema) * long_per + long_ema;
        let out = short_ema - long_ema;
        apo[apo_idx] = out;
        apo_idx += 1;
    }

    // Assertion check
    assert!((apo.as_ptr() as usize - outputs[0].as_ptr() as usize) / std::mem::size_of::<f64>()
        == (size - start) as usize);

    0
}
